rpc/rpc.go

```diff
@@ -884,98 +918,60 @@ func (c *Conn) handleReturn(ctx context.Context, ret rpccp.Return, releaseRet ca
 			q.p.Reject(pr.err)
 			q.bootstrapPromise.Fulfill(q.p.Answer().Client())
 			q.p.ReleaseClients()
-			releaseRet()
+			release()
 		})
 	case q.bootstrapPromise == nil && pr.err != nil:
 		// TODO(someday): send unimplemented message back to remote if
 		// pr.unimplemented == true.
 		q.release = func() {}
 		syncutil.Without(&c.mu, func() {
 			q.p.Reject(pr.err)
-			releaseRet()
+			release()
 		})
 	default:
 		m := ret.Message()
 		q.release = func() {
 			clearCapTable(m)
-			releaseRet()
+			release()
 		}
 		syncutil.Without(&c.mu, func() {
 			q.p.Fulfill(pr.result)
 		})
 	}
-	if err := c.tryLockSender(ctx); err != nil {
-		close(q.finishMsgSend)
-		c.mu.Unlock()
-		c.report(annotate(err, "incoming return: send finish"))
-		return nil
-	}
 	c.mu.Unlock()
 
 	// Send disembargoes.  Failing to send one of these just never lifts
 	// the embargo on our side, but doesn't cause a leak.
 	//
 	// TODO(soon): make embargo resolve to error client.
-	for i := range pr.disembargoes {
-		msg, send, release, err := c.transport.NewMessage(ctx)
-		if err != nil {
-			c.report(failedf("incoming return: send disembargo: create message: %w", err))
-			continue
-		}
-		if err := pr.disembargoes[i].buildDisembargo(msg); err != nil {
-			release()
-			c.report(annotate(err, "incoming return"))
-			continue
-		}
-		err = send()
-		release()
-		if err != nil {
-			c.report(failedf("incoming return: send disembargo: %w", err))
-			continue
+	for _, s := range pr.disembargoes {
+		if err := c.sendMessage(ctx, s.buildDisembargo); err != nil {
```

Comment:
You're calling sendMessage from the receive loop here, rather than sendMessageAsync. Maybe move the error reporting into the callback?